group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/topics/CopartitionedTopicsEnforcerTest.java

```diff
@@ -0,0 +1,203 @@
+
+@SuppressWarnings("OptionalGetWithoutIsPresent")
+public class CopartitionedTopicsEnforcerTest {
+
+    private static final LogContext LOG_CONTEXT = new LogContext();
+
+    private static Integer emptyTopicPartitionProvider(String topic) {
+        return null;
+    }
+
+    private static Integer firstSecondTopicConsistent(String topic) {
+        if (topic.equals("first") || topic.equals("second")) return 2;
+        return null;
+    }
+
+    private static Integer firstSecondTopicInconsistent(String topic) {
+        if (topic.equals("first")) return 2;
+        if (topic.equals("second")) return 1;
+        return null;
+    }
+
+    @Test
+    public void shouldThrowStreamsInconsistentInternalTopicsExceptionIfNoPartitionsFoundForCoPartitionedTopic() {
+        final CopartitionedTopicsEnforcer validator = new CopartitionedTopicsEnforcer(LOG_CONTEXT, CopartitionedTopicsEnforcerTest::emptyTopicPartitionProvider);
+        assertThrows(StreamsInvalidTopologyException.class, () -> validator.enforce(Collections.singleton("topic"),
+            Collections.emptyMap()));
+    }
+
+    @Test
+    public void shouldThrowStreamsInconsistentInternalTopicsExceptionIfPartitionCountsForCoPartitionedTopicsDontMatch() {
+        final CopartitionedTopicsEnforcer validator = new CopartitionedTopicsEnforcer(LOG_CONTEXT, CopartitionedTopicsEnforcerTest::firstSecondTopicInconsistent);
+        assertThrows(StreamsInconsistentInternalTopicsException.class, () -> validator.enforce(Utils.mkSet("first", "second"),
+            Collections.emptyMap()));
+    }
+
+
+    @Test
+    public void shouldEnforceCopartitioningOnRepartitionTopics() {
+        final CopartitionedTopicsEnforcer validator = new CopartitionedTopicsEnforcer(LOG_CONTEXT, CopartitionedTopicsEnforcerTest::firstSecondTopicConsistent);
+        final ConfiguredInternalTopic config = createTopicConfig("repartitioned", 10);
+
+        validator.enforce(Utils.mkSet("first", "second", config.name()),
+            Collections.singletonMap(config.name(), config));
+
+        assertEquals(Optional.of(2), config.numberOfPartitions());
+    }
+
+
+    @Test
+    public void shouldSetNumPartitionsToMaximumPartitionsWhenAllTopicsAreRepartitionTopics() {
+        final CopartitionedTopicsEnforcer validator = new CopartitionedTopicsEnforcer(LOG_CONTEXT, CopartitionedTopicsEnforcerTest::emptyTopicPartitionProvider);
+        final ConfiguredInternalTopic one = createTopicConfig("one", 1);
+        final ConfiguredInternalTopic two = createTopicConfig("two", 15);
+        final ConfiguredInternalTopic three = createTopicConfig("three", 5);
+        final Map<String, ConfiguredInternalTopic> internalTopicConfigs = new HashMap<>();
+
+        internalTopicConfigs.put(one.name(), one);
+        internalTopicConfigs.put(two.name(), two);
+        internalTopicConfigs.put(three.name(), three);
+
+        validator.enforce(Utils.mkSet(one.name(),
+                two.name(),
+                three.name()),
```

Comment:
nit for readability:
```suggestion
        validator.enforce(Utils.mkSet(
            one.name(),
            two.name(),
            three.name()
        ),
```